### PR TITLE
Implement unit testing

### PR DESCRIPTION
**Pull request by chocola-mint**
This PR addresses issue #12, by extracting parsing functionalities into a new class library called `RegAutomation.Core`, and implementing a test suite (`RegAutomation.Core.Tests`) for it. Functions in the `Parser` class in `RegAutomation` have been ported to `ParserBase` in `RegAutomation.Core`, and there is a new generic class `MacroParser<T>` which is inherited by parsers for each reg macro. There is also a parser that inherits directly from `ParserBase`, being `IncludeParser`, which doesn't need to parse macros.

I also made some minor modifications to the implementation of enum parsing and binding (using the enum key directly instead when binding each enum key's value, so constant expressions can be processed automatically without having to evaluate them in C#).

Right now, if you're opening from `solution.sln`, simply using `dotnet test` (or in Visual Studio, "Run All Tests") would trigger `build.bat`, wasting a lot of time running scons before the tests are actually run, so you'd want to run the test project directly instead.

There's a new solution `RegAutomation.sln` that does not contain `example_project` and `ProjectLauncher` that you can use to avoid this issue altogether.

---

I still haven't figured out how to set up GitHub Actions  to automate tests but that can be a future goal. (Though that might be better handled here rather than in my fork) I'm also looking into porting the binding generation code into `Binder` classes so we can test those as well.

### CODE REVIEW

**Comment by haraldwer** [comment]
Good work, as usual!
Just curious, why do we need a new .sln for RegAutomation? 

Notes:

Project and file structure
 - `RegAutomation.sln` `RegAutomation/` `RegAutomation.Core/` and `RegAutomation.Core.Test/` could be moved into a new RegAutomation folder, to keep it all contained in one place. 
 - In `RegAutomation.Core.csproj` and `RegAutomation.Core.Test.csproj`, <OutputPath> should be set to the .temp folder. 
Basically we should try to keep the main folder as clean as possible. 

Code
 - The test classes could be renamed to improve readability, ex `Test_ClassParser`". Similar to how the Pattern classes are named. 
 - Most newly added .cs files have a bunch of unused using directives that should be removed.
 - Maybe add some text about how to set up testing in a readme

**Comment by chocola-mint** [comment]
>Just curious, why do we need a new .sln for RegAutomation?Just curious, why do we need a new .sln for RegAutomation?

In `solution.sln`, running unit tests using Visual Studio's "Run All Tests" command (which corresponds to the CLI command `dotnet test`) would trigger the entire solution's build process before actually running the tests. This means you'll have to spend several minutes waiting for scons to finish building C++ every time you want to run tests, and can be frustrating. This can be side-stepped by running tests in `RegAutomation.Core.Tests` directly (in Visual Studio, right click the Tests project and run tests from the context menu), but it isn't as ergonomical, and this setup wouldn't work from the CLI (as `dotnet test` will always build the solution first before testing), making it harder to automate tests later on.

>In RegAutomation.Core.csproj and RegAutomation.Core.Test.csproj, should be set to the .temp folder.

These aren't auto-generated files, so I don't think it's a good idea to put them into .temp folder.

Aside from that, these all sound like good ideas. I'll take care of them later.

**Comment by chocola-mint** [comment]
Should be all fixed now. Please let me know if there's anything else that needs fixing.

Note that I just included the unit test instructions in the main README as it's really short.

**Comment by haraldwer** [comment]
Nope, looks really good 👏 
One reason for compilation slowdown is that we're not excluding the correct source folders in the SConstruct recursive glob, meaning that the entire folder structure of RegAutomation and ProjectLauncher (including temp files) is included for every compile units. I've made some changes to prevent this.
